ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -345,56 +345,80 @@ module Code_action_error_monoid = struct
+  let dune_actions =
+    Dune.code_actions (State.dune state) params.textDocument.uri
+    |> List.map ~f:(fun a -> `CodeAction a)
   in
-  let+ code_action_results =
-    (* XXX this is a really bad use of resources. we should be batching all the
-       merlin related work *)
-    Fiber.parallel_map ~f:code_action
-      [ Action_destruct.t state
-      ; Action_inferred_intf.t state
-      ; Action_type_annotate.t
-      ; Action_construct.t
-      ; Action_refactor_open.unqualify
-      ; Action_refactor_open.qualify
-      ; Action_add_rec.t
-      ; Action_mark_remove_unused.mark
-      ; Action_mark_remove_unused.remove
-      ]
-  in
-  let code_action_results = List.filter_opt code_action_results in
-  let code_action_results =
-    code_action_results @ Dune.code_actions (State.dune state) doc
-  in
-  match code_action_results with
-  | [] -> None
-  | l -> Some (List.map l ~f:(fun c -> `CodeAction c))
+  match doc with
+  | None -> Fiber.return (Reply.now (Some dune_actions), state)
+  | Some doc -> (
+    match Document.syntax doc with
+    | Ocamllex
+    | Menhir
+    | Cram
+    | Dune ->
+      let state : State.t = Server.state server in
+      Fiber.return (Reply.now (Some dune_actions), state)
+    | Ocaml
+    | Reason ->
+      let reply () =
+        let code_action (ca : Code_action.t) =
+          match params.context.only with
+          | Some set when not (List.mem set ca.kind ~equal:Poly.equal) ->
+            Fiber.return None
+          | Some _
+          | None -> (
+            let+ res =
+              Fiber.map_reduce_errors
+                ~on_error:(fun (exn : Exn_with_backtrace.t) ->
+                  match exn.exn with
+                  | Extend_main.Handshake.Error error ->
+                    Fiber.return (Code_action_error.Need_merlin_extend error)
+                  | _ -> Fiber.return (Code_action_error.Exn exn))
+                (module Code_action_error_monoid)
+                (fun () -> ca.run doc params)
+            in
+            match res with
+            | Ok res -> res
+            | Error Initial -> assert false
+            | Error (Need_merlin_extend _) -> None
+            | Error (Exn exn) -> Exn_with_backtrace.reraise exn)
+        in
+        let+ code_action_results =
+          (* XXX this is a really bad use of resources. we should be batching
+             all the merlin related work *)
+          Fiber.parallel_map ~f:code_action
+            [ Action_destruct.t state
+            ; Action_inferred_intf.t state
+            ; Action_type_annotate.t
+            ; Action_construct.t
+            ; Action_refactor_open.unqualify
+            ; Action_refactor_open.qualify
+            ; Action_add_rec.t
+            ; Action_mark_remove_unused.mark
+            ; Action_mark_remove_unused.remove
+            ]
+        in
+        let code_action_results = List.filter_opt code_action_results in
+        match code_action_results with
+        | [] -> None
```

Comment:
```suggestion
        | [] -> Some dune_actions
```

I think this fixes sometimes missing 'promote' code action

Reply:
Thanks. That's indeed the bug but I fixed in a slightly different way.